nickel/src/capi.rs

```diff
@@ -0,0 +1,1016 @@
+/// function. If that function fails, it will save the error data in your
+/// `nickel_error`.
+pub struct nickel_error {
+    inner: Option<Error>,
+}
+
+/// A Nickel expression.
+///
+/// This might be fully evaluated (for example, if you got it from [`nickel_context_eval_deep`])
+/// or might have unevaluated sub-expressions (if you got it from [`nickel_context_eval_shallow`]).
+// This is only used to hide `Expr`, and to attach C-API-specific documentation to.
+// We only ever use pointers to `nickel_expr`, which are secretly pointers to `Expr`.
+pub struct nickel_expr {}
+
+impl nickel_expr {
+    /// Convert a *nickel_expr to a &Expr.
+    ///
+    /// # Safety
+    ///
+    /// Assumes that `this` was originally a valid pointer to Expr.
+    unsafe fn as_rust(this: &*const Self) -> &Expr {
+        (*this as *const Expr).as_ref().unwrap()
+    }
+
+    /// Convert a *mut nickel_expr to a &mut Expr.
+    ///
+    /// # Safety
+    ///
+    /// Assumes that `this` was originally a valid pointer to Expr.
+    unsafe fn as_rust_mut(this: &mut *mut Self) -> &mut Expr {
+        (*this as *mut Expr).as_mut().unwrap()
+    }
+}
+
+/// A Nickel array.
+///
+/// See [`nickel_expr_is_array`] and [`nickel_expr_as_array`].
+// This is not really an empty struct, it's just an opaque type to
+// hide the implementation. The C API only ever uses `*const Array`,
+// and they are secretly pointers to `nickel_lang_core::term::array::Array`.
+pub struct nickel_array {}
+
+impl nickel_array {
+    unsafe fn as_rust(this: &*const Self) -> Array<'_> {
+        Array {
+            array: (*this as *const nickel_lang_core::term::array::Array)
+                .as_ref()
+                .unwrap(),
+        }
+    }
+}
+
+impl<'a> From<Array<'a>> for *const nickel_array {
+    fn from(arr: Array<'a>) -> Self {
+        arr.array as *const _ as *const nickel_array
+    }
+}
+
+/// A Nickel record.
+///
+/// See [`nickel_expr_is_record`] and [`nickel_expr_as_record`].
+// This is not really an empty struct, it's just an opaque type to
+// hide the implementation. The C API only ever uses `*const Record`,
+// and they are secretly pointers to `nickel_lang_core::term::record::RecordData`.
+pub struct nickel_record {}
+
+impl nickel_record {
+    unsafe fn as_rust(this: &*const Self) -> Record<'_> {
+        Record {
+            data: (*this as *const nickel_lang_core::term::record::RecordData)
+                .as_ref()
+                .unwrap(),
+        }
+    }
+}
+
+impl<'a> From<Record<'a>> for *const nickel_record {
+    fn from(rec: Record<'a>) -> Self {
+        rec.data as *const _ as *const nickel_record
+    }
+}
+
+/// A Nickel string.
+// It would be nice to put `repr(transparent)` here, but (1) we don't
+// actually need to cast `String` pointers to `nickel_string` pointers,
+// and (2) adding `repr(transparent)` makes cbindgen expose `String`
+// even though it's private.
+pub struct nickel_string {
+    inner: String,
```

Comment:
by the way, do you know why you don't have the "expose inner" issue with `String`?

Reply:
Sorry, I think I'm probably just missing something. But why again can't we use that to implement opaque pointers to e.g. `Expr`? What would be wrong with:

```
// (in capi.rs)
pub struct nickel_expr { inner: Expr }
```

Reply:
Is the lifetime the issue? I guess you could do `struct nickel_array { inner: Array<'static> }` (in some sense, we do stop tracking the lifetime of the inner `Array` as soon as it crosses the FFI boundary). It would require a transmute at allocation time, but no (unsafe) cast at usage. Honestly, I'm not entirely sure it's worth the trouble - it's just an idea that came up but I'll let be the judge of what's best.